File: 백준/Platinum/11003. 최솟값 찾기/최솟값 찾기.cc
#include <bits/stdc++.h>
using namespace std;
#define ll long long
#define X first
#define Y second

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    int n, m;
    deque<pair<int, int>> dq;

    cin >> n >> m;

    for (int i = 0; i < n; ++i) {
        int num;
        cin >> num;

        // new number is always the biggest number in deque
        // deque.front number is the smallest number in deque
        // so, if new number is smaller than deque.back number do deque.pop_back
        while (!dq.empty() && dq.back().second >= num) {
            dq.pop_back();
        }
        dq.push_back({i, num});

        // if dq.front number's index is out of range do d.pop_front
        if (dq.front().first <= i - m) {
            dq.pop_front();
        }

        cout << dq.front().second << " ";
    }

  return 0;
}

File: 백준/Platinum/11003. 최솟값 찾기/최솟값 찾기.py
import sys
from collections import deque
r = sys.stdin.readline
sys.setrecursionlimit(10**6)

def solution():
    n, m = map(int, r().split())
    nums = list(map(int, r().split()))
    dq = deque()
    
    for i in range(n):
        # new number is always the biggest number in deque
        # deque.front number is the smallest number in deque
        # so, if new number is smaller than deque.back number do deque.pop_back
        while dq and dq[-1][1] >= nums[i]: dq.pop()
        dq.append((i, nums[i]))
        
        # if dq.front number's index is out of range do d.pop_front
        if dq[0][0] <= i - m: dq.popleft()
        
        print(dq[0][1], end=' ')
        
solution()